src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +134,179 @@ private void logTargetPath() {
+
+        return estimatedPose.plus(odometryPoseToSamplePoseTransform);
+    }
+
+    private Transform2d calculatePoseStandardDeviations(Pose2d estimatedPoseAtObservationTime, PoseEstimatorConstants.StandardDeviations cameraStandardDeviations) {
+        final Transform2d poseEstimateAtObservationTimeToObservationPose = new Transform2d(estimatedPoseAtObservationTime, estimatedPose);
+        final PoseEstimatorConstants.StandardDeviations estimatedPoseStandardDeviations = combineOdometryAndVisionStandardDeviations(cameraStandardDeviations);
+        return scaleTransformFromStandardDeviations(poseEstimateAtObservationTimeToObservationPose, estimatedPoseStandardDeviations);
+    }
+
+    private PoseEstimatorConstants.StandardDeviations combineOdometryAndVisionStandardDeviations(PoseEstimatorConstants.StandardDeviations observationStandardDeviation) {
+        final PoseEstimatorConstants.StandardDeviations odometryStandardDeviations = PoseEstimatorConstants.ODOMETRY_STANDARD_DEVIATIONS;
+
+        return new PoseEstimatorConstants.StandardDeviations(
+                combineOdometryAndVisionStandardDeviation(odometryStandardDeviations.translation(), observationStandardDeviation.translation()),
+                combineOdometryAndVisionStandardDeviation(odometryStandardDeviations.theta(), observationStandardDeviation.theta())
+        );
+    }
+
+    /**
+     * Combines a standard deviation value of the odometry and of the vision result to get a new standard deviation.
+     *
+     * @param odometryStandardDeviation a standard deviation value of the estimated odometry pose
+     * @param visionStandardDeviation   a standard deviation value of the estimated vision pose
+     * @return the combined standard deviation
+     */
+    private double combineOdometryAndVisionStandardDeviation(double odometryStandardDeviation, double visionStandardDeviation) {
+        return odometryStandardDeviation / (odometryStandardDeviation + Math.sqrt(odometryStandardDeviation * visionStandardDeviation));
+    }
+
+    private Transform2d scaleTransformFromStandardDeviations(Transform2d transform, PoseEstimatorConstants.StandardDeviations standardDeviations) {
+        return new Transform2d(
+                transform.getX() * standardDeviations.translation(),
+                transform.getY() * standardDeviations.translation(),
+                transform.getRotation().times(standardDeviations.theta())
+        );
+    }
+
+    private Twist2d calculateNewOdometryPoseDifference(SwerveModulePosition[] swerveModulePositions, Rotation2d gyroHeading) {
+        final Twist2d odometryDifferenceTwist2d = SwerveConstants.KINEMATICS.toTwist2d(lastSwerveModulePositions, swerveModulePositions);
+        return new Twist2d(odometryDifferenceTwist2d.dx, odometryDifferenceTwist2d.dy, gyroHeading.minus(lastGyroHeading).getRadians());
+    }
+
+    private void updateOdometryPositions(SwerveModulePosition[] swerveModulePositions, Rotation2d gyroHeading) {
```

Comment:
I thought we weren't using odometry anymore

Reply:
Why wouldn't we use odometry?